crates/wasi-common/src/fdpool.rs

```diff
@@ -0,0 +1,112 @@
+use std::collections::HashSet;
+use std::fmt;
+use std::hash::Hash;
+
+pub(crate) trait Fd: Copy + Clone + Hash + PartialEq + Eq + Default {
+    fn next(&self) -> Option<Self>;
+}
+
+pub(crate) struct FdPool<T: Fd> {
+    next_alloc: Option<T>,
+    available: Vec<T>,
+    claimed: HashSet<T>,
+}
+
+impl<T: Fd + fmt::Debug> fmt::Debug for FdPool<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("FdPool")
+            .field("next_alloc", &self.next_alloc)
+            .field("available", &self.available)
+            .field("claimed", &self.claimed)
+            .finish()
+    }
+}
+
+impl<T: Fd> FdPool<T> {
+    pub fn new() -> Self {
+        Self {
+            next_alloc: Some(T::default()),
+            available: Vec::new(),
+            claimed: HashSet::new(),
+        }
+    }
+
+    pub fn allocate(&mut self) -> Option<T> {
+        if let Some(fd) = self.available.pop() {
+            // Since we've had free, unclaimed handle in the pool,
+            // simply claim it and return.
+            self.claimed.insert(fd);
+            return Some(fd);
+        }
+        // There are no free handles available in the pool, so try
+        // allocating an additional one into the pool. If we've
+        // reached our max number of handles, we will fail with None
+        // instead.
+        let fd = match self.next_alloc.take() {
```

Comment:
This could use `?` I think if you wanted to be super nifty

Reply:
Of course, well spotted, thanks!